R/data.table.R

```diff
@@ -1294,7 +1296,7 @@ replace_order = function(isub, verbose, env) {
           identical(irows, integer(0L)) && !bynull,
           length(irows) && !anyNA(irows) && all(irows==0L) ## anyNA() because all() returns NA (not FALSE) when irows is all-NA. TODO: any way to not check all 'irows' values?
           ))
-        if (is.atomic(jval)) jval = jval[0L] else jval = lapply(jval, `[`, 0L)
+        jval = lapply(jval, `[`, 0L)
```

Comment:
Just above you can see this branch enters with either `is.list(jval)` or `!missingby`:

https://github.com/Rdatatable/data.table/pull/3761/files#diff-650e4a11ed4384f8e560a6ebfff4ff53L1287

```
if ((is.call(jsub) && is.list(jval) && jsub[[1L]] != "get" && !is.object(jval)) || !missingby)
```

 Obviously `is.atomic(jval)` is impossible in the first case. You have to scroll a bit further up but I think the second case is also impossible:

https://github.com/Rdatatable/data.table/pull/3761/files#diff-650e4a11ed4384f8e560a6ebfff4ff53L1168

```
if (missingby || bynull || (!byjoin && !length(byval)))
```
So we've either got `bynull` or `(!byjoin && !length(byval))`. But I'm not sure how `length(byval)` can be `0` but `bynull` is `FALSE` -- e.g. `by = integer()` would have `byval = list(integer = integer())`, so it's already en-`list`ed.